CI: dead code removal check

yvan-sraka opened the pull request:
This PR adds a CI job that runs the nix run `github:astro/deadnix -- --edit --no-lambda-pattern-names` command to remove dead code. If this command results in any uncommitted changes, the CI job will fail, indicating the presence of dead code in the repository that needs to be removed. See https://github.com/input-output-hk/haskell.nix/pull/1756

angerman commented:
Can this create a patch? Or even better suggestions to commit?

yvan-sraka commented:
Yes! We can use GitHub Actions like, e.g., [`reviewdog/action-suggester`](https://github.com/reviewdog/action-suggester) to generate suggestions :) WDYT?

yvan-sraka commented:
Not sure to understand if that's the GA which is faulty, otherwise I would gladly use it also to check Nix code formatting / LSP insights too :)

I'm not sure if [the ReviewDog GA is faulty](https://github.com/input-output-hk/haskell.nix/actions/runs/5308697477/jobs/9608457749?pr=1983#step:4:27)? Otherwise, I would gladly use it to check Nix code formatting and display LSP insights as well :)

andreabedini commented on .github/workflows/pipeline.yml:
```diff
@@ -240,3 +240,12 @@ jobs:
+    steps:
+      - uses: actions/checkout@v3
+      - run: nix run github:astro/deadnix -- --edit --no-lambda-pattern-names
+      - uses: yvan-sraka/action-suggester@master
```
Is this really needed? Notice that you are running nix in the second step and then installing it again as part of the third step.

yvan-sraka replied:
The [PATCH I applied in my fork of the GHA](https://github.com/reviewdog/action-suggester/compare/master...yvan-sraka:action-suggester:master) is because, AFAIK, we're using custom NixOS job runners that would make the action fail otherwise https://github.com/reviewdog/action-suggester/issues/35 …

andreabedini replied:
The issue you link has now been closed. Is your patch still necessary?

andreabedini reviewed (changes requested):
Is it supposed to fail if there are warnings? because it didn't

https://github.com/input-output-hk/haskell.nix/actions/runs/5789319513/job/15690115119#step:3:7

yvan-sraka commented:
> Is it supposed to fail if there are warnings? because it didn't

I choose to not make a CI test fail, but rather to just display hints in the PR diffs view (that's what https://github.com/reviewdog/action-suggester does). I'm not sure if it's more convenient?

andreabedini reviewed (changes requested):
Can you make a separate workflow? `pipeline.yml` is a bit of a kitchen sink. Maybe move all lints to a new `lints.yml` workflow?

Is there anywhere I can see a preview of the output?

yvan-sraka commented:
> Can you make a separate workflow? `pipeline.yml` is a bit of a kitchen sink. Maybe move all lints to a new `lints.yml` workflow?

Yes! I updated the PR :)

> Is there anywhere I can see a preview of the output?

I fixed it, I now directly rely on https://github.com/reviewdog/reviewdog?tab=readme-ov-file#diff (I would prefer to not have to rely on external tooling, but the needed GitHub feature seems missing https://github.com/cli/cli/issues/5905 …)

yvan-sraka commented:
So I give it a try to it in my fork, and here what's expected: https://github.com/yvan-sraka/haskell.nix/pull/1 :)
